File: typescript/src/agents/tools/web-search.ts
/**
 * ``WebSearchTool`` — backed by Tavily, Brave, or Amazon Bedrock AgentCore.
 *
 * Returns a LangChain.js ``BaseTool`` whose output flows through Compresr's
 * ``compresrToolMiddleware`` automatically when used with the SDK's engine
 * (``client.messages.create`` / ``client.chat.completions.create`` /
 * ``client.run``).
 *
 * Provider-native server tools (Anthropic ``web_search_20250305``, OpenAI
 * ``web_search_preview``, Gemini ``google_search``) are intentionally NOT
 * supported here — they execute server-side and return opaque/encrypted
 * content that Compresr cannot read or compress. Use a real search API
 * (Tavily or Brave) so the result is plaintext we can compress.
 *
 * Mirrors Python ``compresr/agents/tools/web_search.py``.
 */
import type { StructuredToolInterface } from '@langchain/core/tools';

import { CompresrError } from '../../errors/index.js';
import {
  buildAgentCoreClient,
  type AgentCoreConfig,
} from './_agentcore.js';

/** Public return type for the web-search factories — any tool that satisfies
 * LangChain.js's ``StructuredToolInterface`` is acceptable. We use the
 * interface form (not the abstract class) so we avoid generic parameter
 * gymnastics at the call site. */
export type WebSearchToolInstance = StructuredToolInterface;

export interface TavilyOptions {
  apiKey?: string;
  maxResults?: number;
  allowedDomains?: ReadonlyArray<string>;
  blockedDomains?: ReadonlyArray<string>;
  /** Additional kwargs forwarded to the underlying LangChain tool. */
  extra?: Record<string, unknown>;
}

export interface BraveOptions {
  apiKey?: string;
  maxResults?: number;
  /** Additional kwargs forwarded to the underlying LangChain tool. */
  extra?: Record<string, unknown>;
}

/**
 * Options for ``WebSearchTool.agentcore`` (Amazon Bedrock AgentCore web
 * search). Each gateway field is resolved with precedence
 * **explicit option → environment variable**:
 *
 * | Field          | Env var (fallback)                                |
 * |----------------|---------------------------------------------------|
 * | gatewayUrl     | AGENTCORE_GATEWAY_MCP_URL (GATEWAY_MCP_URL)       |
 * | cognitoTokenUrl| AGENTCORE_COGNITO_TOKEN_URL (COGNITO_TOKEN_URL)   |
 * | clientId       | AGENTCORE_COGNITO_CLIENT_ID (COGNITO_CLIENT_ID)   |
 * | clientSecret   | AGENTCORE_COGNITO_CLIENT_SECRET (COGNITO_CLIENT_SECRET) |
 * | scope          | AGENTCORE_COGNITO_SCOPE (COGNITO_SCOPE)           |
 *
 * ``allowedDomains`` / ``blockedDomains`` are accepted for signature parity
 * with Tavily/Brave but are **unsupported** by AgentCore web search (no native
 * domain filter) — use Tavily if you need domain filtering.
 */
export interface AgentCoreOptions {
  gatewayUrl?: string;
  cognitoTokenUrl?: string;
  clientId?: string;
  clientSecret?: string;
  scope?: string;
  maxResults?: number;
  /** Unsupported by AgentCore — accepted for parity, otherwise ignored. */
  allowedDomains?: ReadonlyArray<string>;
  /** Unsupported by AgentCore — accepted for parity, otherwise ignored. */
  blockedDomains?: ReadonlyArray<string>;
  /** Reserved for forward compatibility; currently unused. */
  extra?: Record<string, unknown>;
}

/** Flatten ``{results: [{title, url, content}, ...]}`` into blank-line
 * -separated plain-text blocks. ``latte_v1`` no-ops on JSON-shaped input,
 * so this reshape is what makes Tavily output compressible by the middleware. */
function flattenSearchResults(out: unknown): string {
  if (typeof out === 'string') return out;
  if (out && typeof out === 'object' && 'results' in out) {
    const results = (out as { results?: unknown }).results;
    if (Array.isArray(results)) {
      const parts: string[] = [];
      for (const r of results) {
        if (!r || typeof r !== 'object') continue;
        const rr = r as Record<string, unknown>;
        const title = typeof rr['title'] === 'string' ? rr['title'].trim() : '';
        const url = typeof rr['url'] === 'string' ? rr['url'].trim() : '';
        const content =
          typeof rr['content'] === 'string'
            ? rr['content'].trim()
            : typeof rr['snippet'] === 'string'
              ? rr['snippet'].trim()
              : '';
        const block = [title, url, content].filter(Boolean).join('\n');
        if (block) parts.push(block);
      }
      if (parts.length > 0) return parts.join('\n\n');
    }
  }
  return JSON.stringify(out);
}

async function buildTavily(options: TavilyOptions): Promise<WebSearchToolInstance> {
  let mod: Record<string, unknown>;
  try {
    mod = await import('@langchain/tavily');
  } catch {
    throw new CompresrError(
      "WebSearchTool.tavily requires '@langchain/tavily'. " +
        'Install with: npm install @langchain/tavily',
      'missing_peer_dependency'
    );
  }
  const TavilySearch = (mod['TavilySearch'] ?? mod['default']) as
    | (new (init: Record<string, unknown>) => unknown)
    | undefined;
  if (typeof TavilySearch !== 'function') {
    throw new CompresrError(
      "'@langchain/tavily' is missing TavilySearch — upgrade the package.",
      'missing_peer_dependency'
    );
  }
  const init: Record<string, unknown> = {
    maxResults: options.maxResults ?? 5,
  };
  if (options.apiKey) {
    init['tavilyApiKey'] = options.apiKey;
  }
  if (options.allowedDomains !== undefined) {
    init['includeDomains'] = [...options.allowedDomains];
  }
  if (options.blockedDomains !== undefined) {
    init['excludeDomains'] = [...options.blockedDomains];
  }
  // Merge ``extra`` first so trusted fields (apiKey, includeDomains, etc.)
  // win on collision — see audit TS-M3.
  const merged = { ...options.extra, ...init };
  const base = new TavilySearch(merged) as {
    invoke: (args: { query: string }) => Promise<unknown>;
  };

  // Wrap with a tool() so the returned ToolMessage content is plain text
  // (latte_v1 no-ops on JSON), and so an explicit ``query`` schema reaches
  // the LLM (parity with brave).
  const { tool } = (await import('@langchain/core/tools')) as {
    tool: (
      fn: (input: { query: string }) => Promise<string>,
      meta: { name: string; description: string; schema: unknown }
    ) => WebSearchToolInstance;
  };
  const { z } = (await import('zod')) as { z: typeof import('zod').z };

  return tool(
    async ({ query }) => flattenSearchResults(await base.invoke({ query })),
    {
      name: 'tavily_search',
      description:
        'Tavily web search. Use to answer questions about current events.',
      schema: z.object({
        query: z.string().max(2000).describe('The search query string.'),
      }),
    }
  );
}

async function buildBrave(options: BraveOptions): Promise<WebSearchToolInstance> {
  let mod: Record<string, unknown>;
  try {
    mod = (await import(
      // @ts-expect-error — optional peer; types not guaranteed at compile time
      '@langchain/community/tools/brave_search'
    )) as Record<string, unknown>;
  } catch {
    throw new CompresrError(
      "WebSearchTool.brave requires '@langchain/community'. " +
        'Install with: npm install @langchain/community',
      'missing_peer_dependency'
    );
  }
  const BraveSearch = (mod['BraveSearch'] ?? mod['default']) as
    | (new (init: Record<string, unknown>) => unknown)
    | undefined;
  if (typeof BraveSearch !== 'function') {
    throw new CompresrError(
      "'@langchain/community' is missing BraveSearch — upgrade the package.",
      'missing_peer_dependency'
    );
  }
  const key =
    options.apiKey ??
    process.env['BRAVE_SEARCH_API_KEY'] ??
    process.env['BRAVE_API_KEY'];
  if (!key) {
    throw new CompresrError(
      'Brave requires apiKey or BRAVE_SEARCH_API_KEY env var.',
      'missing_api_key'
    );
  }
  const init: Record<string, unknown> = {
    apiKey: key,
    searchKwargs: { count: options.maxResults ?? 5 },
  };
  // Merge ``extra`` first so trusted fields (apiKey, searchKwargs) win on
  // collision — see audit TS-M3.
  const merged = { ...options.extra, ...init };
  return new BraveSearch(merged) as WebSearchToolInstance;
}

/** Clamp the requested result count into AgentCore's supported 1–25 range. */
function clampMaxResults(value: number | undefined): number {
  const n = value ?? 5;
  if (!Number.isFinite(n)) return 5;
  return Math.min(25, Math.max(1, Math.trunc(n)));
}

/** Each AgentCore config field and the env vars it falls back to. */
const AGENTCORE_ENV: ReadonlyArray<{
  key: keyof Omit<AgentCoreConfig, 'maxResults'>;
  optionLabel: string;
  envVars: ReadonlyArray<string>;
}> = [
  {
    key: 'gatewayUrl',
    optionLabel: 'gatewayUrl',
    envVars: ['AGENTCORE_GATEWAY_MCP_URL', 'GATEWAY_MCP_URL'],
  },
  {
    key: 'cognitoTokenUrl',
    optionLabel: 'cognitoTokenUrl',
    envVars: ['AGENTCORE_COGNITO_TOKEN_URL', 'COGNITO_TOKEN_URL'],
  },
  {
    key: 'clientId',
    optionLabel: 'clientId',
    envVars: ['AGENTCORE_COGNITO_CLIENT_ID', 'COGNITO_CLIENT_ID'],
  },
  {
    key: 'clientSecret',
    optionLabel: 'clientSecret',
    envVars: ['AGENTCORE_COGNITO_CLIENT_SECRET', 'COGNITO_CLIENT_SECRET'],
  },
  {
    key: 'scope',
    optionLabel: 'scope',
    envVars: ['AGENTCORE_COGNITO_SCOPE', 'COGNITO_SCOPE'],
  },
];

/** URL-valued config fields that must use ``https://`` to avoid leaking the
 * Cognito client secret / bearer token in cleartext. */
const AGENTCORE_URL_FIELDS: ReadonlyArray<keyof Omit<AgentCoreConfig, 'maxResults'>> =
  ['gatewayUrl', 'cognitoTokenUrl'];

/** Resolve config with precedence explicit option → env var; collect misses. */
function resolveAgentCoreConfig(options: AgentCoreOptions): AgentCoreConfig {
  const resolved: Record<string, string> = {};
  const missing: string[] = [];
  for (const { key, optionLabel, envVars } of AGENTCORE_ENV) {
    const fromOption = options[key];
    let value: string | undefined =
      typeof fromOption === 'string' && fromOption.length > 0
        ? fromOption
        : undefined;
    if (value === undefined) {
      for (const envVar of envVars) {
        const candidate = process.env[envVar];
        if (candidate) {
          value = candidate;
          break;
        }
      }
    }
    if (value === undefined) {
      missing.push(`${optionLabel} (env: ${envVars.join(' or ')})`);
    } else {
      resolved[key] = value;
    }
  }
  if (missing.length > 0) {
    throw new CompresrError(
      'WebSearchTool.agentcore is missing required config: ' +
        missing.join('; ') +
        '.',
      'missing_config'
    );
  }
  for (const field of AGENTCORE_URL_FIELDS) {
    if (!resolved[field].startsWith('https://')) {
      throw new CompresrError(
        `WebSearchTool.agentcore: '${field}' must use https:// ` +
          '(refusing to send credentials in cleartext).',
        'invalid_config'
      );
    }
  }
  return {
    gatewayUrl: resolved['gatewayUrl'],
    cognitoTokenUrl: resolved['cognitoTokenUrl'],
    clientId: resolved['clientId'],
    clientSecret: resolved['clientSecret'],
    scope: resolved['scope'],
    maxResults: clampMaxResults(options.maxResults),
  };
}

async function buildAgentCore(
  options: AgentCoreOptions
): Promise<WebSearchToolInstance> {
  const config = resolveAgentCoreConfig(options);
  const client = buildAgentCoreClient(config);

  const { tool } = (await import('@langchain/core/tools')) as {
    tool: (
      fn: (input: { query: string }) => Promise<string>,
      meta: { name: string; description: string; schema: unknown }
    ) => WebSearchToolInstance;
  };
  const { z } = (await import('zod')) as { z: typeof import('zod').z };

  return tool(
    async ({ query }) =>
      flattenSearchResults({
        results: await client.search(query, config.maxResults),
      }),
    {
      name: 'agentcore_web_search',
      description:
        'Amazon Bedrock AgentCore web search. Use to answer questions ' +
        'about current events.',
      schema: z.object({
        query: z.string().max(2000).describe('The search query string.'),
      }),
    }
  );
}

/**
 * Factory namespace for web-search tools — pick a provider via
 * ``WebSearchTool.tavily(...)``, ``WebSearchTool.brave(...)``, or
 * ``WebSearchTool.agentcore(...)``.
 *
 * Each method returns a LangChain.js ``StructuredToolInterface`` ready to
 * pass into the engine's ``tools: []`` array. The interface form is used
 * (instead of the abstract ``StructuredTool`` class) so callers can drop
 * in any compatible tool implementation without hitting generic-parameter
 * mismatches.
 */
export const WebSearchTool = {
  tavily(options: TavilyOptions = {}): Promise<WebSearchToolInstance> {
    return buildTavily(options);
  },
  brave(options: BraveOptions = {}): Promise<WebSearchToolInstance> {
    return buildBrave(options);
  },
  agentcore(options: AgentCoreOptions = {}): Promise<WebSearchToolInstance> {
    return buildAgentCore(options);
  },
} as const;

export type WebSearchTool = typeof WebSearchTool;

/** Direct factory for the Tavily variant — preferred when ``tools: [...]``
 * needs a synchronous reference. Returns a promise that resolves to a tool. */
export function createWebSearchTool(
  provider: 'tavily',
  options?: TavilyOptions
): Promise<WebSearchToolInstance>;
export function createWebSearchTool(
  provider: 'brave',
  options?: BraveOptions
): Promise<WebSearchToolInstance>;
export function createWebSearchTool(
  provider: 'agentcore',
  options?: AgentCoreOptions
): Promise<WebSearchToolInstance>;
export function createWebSearchTool(
  provider: 'tavily' | 'brave' | 'agentcore',
  options: TavilyOptions | BraveOptions | AgentCoreOptions = {}
): Promise<WebSearchToolInstance> {
  if (provider === 'tavily') return buildTavily(options);
  if (provider === 'brave') return buildBrave(options);
  if (provider === 'agentcore') return buildAgentCore(options);
  throw new CompresrError(
    `Unknown web search provider: '${String(provider)}'. ` +
      "Use 'tavily', 'brave', or 'agentcore'.",
    'invalid_provider'
  );
}

File: python/compresr/agents/tools/web_search.py
"""WebSearchTool — backed by Tavily, Brave, or Amazon Bedrock AgentCore.

Returns a LangChain ``BaseTool`` whose output flows through Compresr's
``CompresrToolMiddleware`` automatically when used with the SDK's engine
(``client.messages.create`` / ``client.chat.completions.create`` /
``client.run``).

The ``agentcore`` provider reaches Amazon Bedrock AgentCore web search via a
Cognito OAuth handshake + MCP streamable-HTTP session; its runtime deps are
optional (``pip install compresr[agentcore]``).

Provider-native server tools (Anthropic ``web_search_20250305``, OpenAI
``web_search_preview``, Gemini ``google_search``) are intentionally NOT
supported here — they execute server-side and return opaque/encrypted
content that Compresr cannot read or compress. Use a real search API
(Tavily or Brave) so the result is plaintext we can compress.
"""

from __future__ import annotations

import os
from typing import Any, Optional


class WebSearchTool:
    """Factory for a LangChain web-search tool. Default provider is Tavily.

    Args:
        provider: ``"tavily"`` (default) or ``"brave"``.
        api_key: Provider API key. Falls back to ``TAVILY_API_KEY`` /
            ``BRAVE_SEARCH_API_KEY`` env vars.
        max_results: How many results to return.
        allowed_domains: Restrict to these domains (Tavily: ``include_domains``).
        blocked_domains: Exclude these domains (Tavily: ``exclude_domains``).
        **extra: Forwarded to the underlying LangChain tool constructor.

    Returns:
        A ``langchain_core.tools.BaseTool`` ready to pass to
        ``client.messages.create(tools=[...])``.

    Prefer the provider-named classmethod factories for discoverability
    and static-typing friendliness::

        WebSearchTool.tavily(api_key=..., max_results=5)
        WebSearchTool.brave(api_key=..., max_results=5)
        WebSearchTool.agentcore(gateway_url=..., ...)
    """

    @classmethod
    def tavily(
        cls,
        *,
        api_key: Optional[str] = None,
        max_results: int = 5,
        allowed_domains: Optional[list[str]] = None,
        blocked_domains: Optional[list[str]] = None,
        **extra: Any,
    ) -> Any:
        """Tavily-backed web search. Returns a LangChain ``BaseTool``.

        Compresr's middleware compresses the tool output automatically.
        """
        return cls(
            provider="tavily",
            api_key=api_key,
            max_results=max_results,
            allowed_domains=allowed_domains,
            blocked_domains=blocked_domains,
            **extra,
        )

    @classmethod
    def brave(
        cls,
        *,
        api_key: Optional[str] = None,
        max_results: int = 5,
        **extra: Any,
    ) -> Any:
        """Brave-backed web search. Returns a LangChain ``BaseTool``."""
        return cls(
            provider="brave",
            api_key=api_key,
            max_results=max_results,
            **extra,
        )

    @classmethod
    def agentcore(
        cls,
        *,
        gateway_url: Optional[str] = None,
        cognito_token_url: Optional[str] = None,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None,
        scope: Optional[str] = None,
        max_results: int = 5,
        allowed_domains: Optional[list[str]] = None,
        blocked_domains: Optional[list[str]] = None,
        **extra: Any,
    ) -> Any:
        """Amazon Bedrock AgentCore web search. Returns a LangChain ``BaseTool``.

        Config resolves from explicit args first, then env vars
        (``AGENTCORE_GATEWAY_MCP_URL`` / ``GATEWAY_MCP_URL``,
        ``AGENTCORE_COGNITO_TOKEN_URL`` / ``COGNITO_TOKEN_URL``,
        ``AGENTCORE_COGNITO_CLIENT_ID`` / ``COGNITO_CLIENT_ID``,
        ``AGENTCORE_COGNITO_CLIENT_SECRET`` / ``COGNITO_CLIENT_SECRET``,
        ``AGENTCORE_COGNITO_SCOPE`` / ``COGNITO_SCOPE``). Requires the
        ``agentcore`` extra: ``pip install compresr[agentcore]``.
        """
        return cls(
            provider="agentcore",
            gateway_url=gateway_url,
            cognito_token_url=cognito_token_url,
            client_id=client_id,
            client_secret=client_secret,
            scope=scope,
            max_results=max_results,
            allowed_domains=allowed_domains,
            blocked_domains=blocked_domains,
            **extra,
        )

    def __new__(
        cls,
        *,
        provider: str = "tavily",
        api_key: Optional[str] = None,
        max_results: int = 5,
        allowed_domains: Optional[list[str]] = None,
        blocked_domains: Optional[list[str]] = None,
        gateway_url: Optional[str] = None,
        cognito_token_url: Optional[str] = None,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None,
        scope: Optional[str] = None,
        **extra: Any,
    ) -> Any:
        if provider == "tavily":
            return _build_tavily(
                api_key=api_key,
                max_results=max_results,
                allowed_domains=allowed_domains,
                blocked_domains=blocked_domains,
                **extra,
            )
        if provider == "brave":
            return _build_brave(
                api_key=api_key,
                max_results=max_results,
                allowed_domains=allowed_domains,
                blocked_domains=blocked_domains,
                **extra,
            )
        if provider == "agentcore":
            return _build_agentcore(
                gateway_url=gateway_url,
                cognito_token_url=cognito_token_url,
                client_id=client_id,
                client_secret=client_secret,
                scope=scope,
                max_results=max_results,
                allowed_domains=allowed_domains,
                blocked_domains=blocked_domains,
                **extra,
            )
        raise ValueError(
            f"Unknown web search provider: {provider!r}. Use 'tavily', 'brave', or 'agentcore'."
        )


def _flatten_search_results(out: Any) -> str:
    """Flatten ``{"results": [{title, url, content}, ...]}`` into blank-line
    -separated plain-text blocks.
    """
    if isinstance(out, str):
        return out
    if isinstance(out, dict):
        results = out.get("results")
        if isinstance(results, list):
            parts: list[str] = []
            for r in results:
                if not isinstance(r, dict):
                    continue
                title = (r.get("title") or "").strip()
                url = (r.get("url") or "").strip()
                content = (r.get("content") or r.get("snippet") or "").strip()
                block = "\n".join(p for p in (title, url, content) if p)
                if block:
                    parts.append(block)
            if parts:
                return "\n\n".join(parts)
    import json

    return json.dumps(out, ensure_ascii=False, default=str)


def _build_tavily(
    *,
    api_key: Optional[str],
    max_results: int,
    allowed_domains: Optional[list[str]],
    blocked_domains: Optional[list[str]],
    **extra: Any,
) -> Any:
    try:
        from langchain_tavily import TavilySearch
    except ImportError as exc:
        raise ImportError(
            "WebSearchTool(provider='tavily') requires langchain-tavily. "
            "Install with: pip install compresr[agents-tavily]"
        ) from exc
    key = api_key or os.environ.get("TAVILY_API_KEY")
    if not key:
        raise ValueError("Tavily requires api_key= or TAVILY_API_KEY env var.")
    kwargs: dict = {"max_results": max_results, "tavily_api_key": key}
    if allowed_domains is not None:
        kwargs["include_domains"] = list(allowed_domains)
    if blocked_domains is not None:
        kwargs["exclude_domains"] = list(blocked_domains)
    kwargs.update(extra)
    base = TavilySearch(**kwargs)

    # Wrap with a StructuredTool so the returned ToolMessage content is plain
    # text (latte_v1 no-ops on JSON), and so an explicit ``query`` schema
    # reaches the LLM (parity with brave).
    from langchain_core.tools import StructuredTool
    from pydantic import BaseModel, Field

    class _TavilySearchInput(BaseModel):
        query: str = Field(..., description="The search query string.")

    def _run(query: str) -> str:
        return _flatten_search_results(base.invoke({"query": query}))

    return StructuredTool.from_function(
        func=_run,
        name="tavily_search",
        description=getattr(base, "description", None)
        or "Tavily web search. Use to answer questions about current events.",
        args_schema=_TavilySearchInput,
    )


def _build_brave(
    *,
    api_key: Optional[str],
    max_results: int,
    allowed_domains: Optional[list[str]],
    blocked_domains: Optional[list[str]],
    **extra: Any,
) -> Any:
    import warnings as _warnings

    try:
        with _warnings.catch_warnings():
            # langchain-community emits a sunset DeprecationWarning. Brave's
            # canonical home is still langchain-community as of this release;
            # suppress that one specific notice instead of every warning.
            _warnings.filterwarnings(
                "ignore",
                category=DeprecationWarning,
                message=r".*langchain-community.*sunset.*",
            )
            from langchain_community.tools import BraveSearch
    except ImportError as exc:
        raise ImportError(
            "WebSearchTool(provider='brave') requires langchain-community. "
            "Install with: pip install compresr[agents-brave]"
        ) from exc
    key = api_key or os.environ.get("BRAVE_SEARCH_API_KEY") or os.environ.get("BRAVE_API_KEY")
    if not key:
        raise ValueError("Brave requires api_key= or BRAVE_SEARCH_API_KEY env var.")
    search_kwargs: dict = {"count": max_results}
    # Brave's allowed/blocked domain support is via Goggles — out of scope for v1.
    if allowed_domains or blocked_domains:
        import warnings

        warnings.warn(
            "Brave does not natively support allowed_domains/blocked_domains; "
            "use Tavily for domain filtering.",
            stacklevel=3,
        )
    search_kwargs.update(extra)
    base = BraveSearch.from_api_key(api_key=key, search_kwargs=search_kwargs)

    # ``langchain_community.tools.BraveSearch`` ships with ``args_schema=None``,
    # so when the LLM (especially Anthropic Claude via tool_use) is asked to
    # call it, the tool schema sent in the prompt has no ``query`` field. The
    # model emits an empty args dict, and ``BraveSearch._run(query)`` then
    # raises ``TypeError: missing 1 required positional argument: 'query'``.
    # Wrap with an explicit pydantic schema so the LLM always sees ``query``.
    from langchain_core.tools import StructuredTool
    from pydantic import BaseModel, Field

    class _BraveSearchInput(BaseModel):
        query: str = Field(..., description="The search query string.")

    return StructuredTool.from_function(
        func=lambda query: base.run(query),
        name="brave_search",
        description=base.description
        or "Brave web search. Use it to answer questions about current events.",
        args_schema=_BraveSearchInput,
    )


# Field -> (explicit-arg resolver, ordered env-var fallbacks). The first env var
# is the AgentCore-namespaced name; the second matches the reference client.
_AGENTCORE_ENV: dict[str, tuple[str, ...]] = {
    "gateway_url": ("AGENTCORE_GATEWAY_MCP_URL", "GATEWAY_MCP_URL"),
    "cognito_token_url": ("AGENTCORE_COGNITO_TOKEN_URL", "COGNITO_TOKEN_URL"),
    "client_id": ("AGENTCORE_COGNITO_CLIENT_ID", "COGNITO_CLIENT_ID"),
    "client_secret": ("AGENTCORE_COGNITO_CLIENT_SECRET", "COGNITO_CLIENT_SECRET"),
    "scope": ("AGENTCORE_COGNITO_SCOPE", "COGNITO_SCOPE"),
}


def _resolve_agentcore_config(values: dict) -> dict:
    """Resolve each AgentCore field via explicit arg → env-var fallbacks.

    Raises ``ValueError`` naming every missing field and its env-var fallbacks.
    Never echoes the resolved secret value.
    """
    resolved: dict = {}
    missing: list[str] = []
    for field, env_names in _AGENTCORE_ENV.items():
        value = values.get(field)
        for env_name in env_names:
            if value:
                break
            value = os.environ.get(env_name)
        if value:
            resolved[field] = value
        else:
            missing.append(f"{field} ({' or '.join(env_names)})")
    if missing:
        raise ValueError(
            "WebSearchTool.agentcore is missing required config: "
            + "; ".join(missing)
            + ". Pass it explicitly or set the env var."
        )
    return resolved


def _build_agentcore(
    *,
    gateway_url: Optional[str],
    cognito_token_url: Optional[str],
    client_id: Optional[str],
    client_secret: Optional[str],
    scope: Optional[str],
    max_results: int,
    allowed_domains: Optional[list[str]],
    blocked_domains: Optional[list[str]],
    **extra: Any,
) -> Any:
    from ._agentcore import AgentCoreClient, AgentCoreConfig

    if allowed_domains or blocked_domains:
        import warnings

        warnings.warn(
            "AgentCore web search does not natively support allowed_domains/"
            "blocked_domains; use Tavily for domain filtering.",
            stacklevel=3,
        )

    resolved = _resolve_agentcore_config(
        {
            "gateway_url": gateway_url,
            "cognito_token_url": cognito_token_url,
            "client_id": client_id,
            "client_secret": client_secret,
            "scope": scope,
        }
    )
    config = AgentCoreConfig(max_results=max_results, **resolved)
    # One shared client so the cached bearer token is reused across calls.
    client = AgentCoreClient(config)

    from langchain_core.tools import StructuredTool
    from pydantic import BaseModel, Field

    class _AgentCoreSearchInput(BaseModel):
        query: str = Field(..., description="The search query string.")

    def _run(query: str) -> str:
        return _flatten_search_results({"results": client.search(query, max_results)})

    tool_kwargs: dict = {
        "func": _run,
        "name": "agentcore_web_search",
        "description": "Amazon Bedrock AgentCore web search. Use to answer "
        "questions about current events.",
        "args_schema": _AgentCoreSearchInput,
    }
    # Unlike tavily/brave, there is no underlying SDK call to forward
    # provider-specific kwargs to — extras are accepted (for signature parity)
    # and passed straight through to StructuredTool.from_function.
    tool_kwargs.update(extra)
    return StructuredTool.from_function(**tool_kwargs)


__all__ = ["WebSearchTool"]
